### ISSUE TITLE
Seemingly spurious CS8619 when calling ToDictionary after recent VS update

### ISSUE THREAD

**Issue by madelson** [dotnet/roslyn#50286]
**Version Used**: 

Visual Studio 16.8.3 compiling against netstandard2.0 with LangVersion 8.0 on Windows 10.

**Steps to Reproduce**:

The following code causes the warning:

```
IReadOnlyDictionary<string, object?> dictionary = new[] { new KeyValuePair<string, object?>("a", null) }
    .ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
```

The warning says:
```
CS8619	Nullability of reference types in value of type 'Dictionary<string, object>' doesn't match target type 'IReadOnlyDictionary<string, object?>'.
```

Changing the value selector to `kvp => (object?)kvp.Value` resolves the warning.

**Expected Behavior**:

I don't think this should emit a warning; the type of kvp is clearly `KeyValuePair<string, object?>`, so `kvp.Value` should be of type `object?` which means it should infer `Dictionary<string, object?>` by itself.

**Actual Behavior**:

Emits warning.

**Comment by karataliu** [comment]
Got same issue in SDK 5.0.100, in SDK 5.0.103/vs16.8.5 issue was gone. Is it a VS issue or SDK issue?

**Comment by RikkiGibson** [comment]
I can repro the issue in all the compilers that *should* have shipped in SDK 5.0.1xx. I can't repro the issue in latest master. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+BiAdgVwBtCJhC4ACOXU8gWAChGABABguYEYAWAbhfZcAdABkAlrgCO/JgwCSAJTgQUAeVyEAngBExAYxhiA9jQSaAPF1ZIKR4ACs4BgPwA+Cin2GTEMxQC8FLhwAO4A2gC6FADeQaEUANJwmgBqEIT4cAAKEGIIlpzWtg5OMG4AFABEEJU2BMQAlBQAvowU7RRCACpGugbGpprlANYAbgAOAe5j40JJmjYzUxQzQmkZcA0yQA==)

@karataliu can you try adding `#error version` to your source code and sharing the diagnostic that is produced.

**Comment by karataliu** [comment]
@RikkiGibson 

failure one:
`error CS8304: Compiler version: '3.8.0-5.20519.18 (4c195c3a)'. Language version: 8.0.`

passing one:
`error CS8304: Compiler version: '3.8.0-5.20604.10 (9ed4b774)'. Language version: 8.0.`

**Comment by RikkiGibson** [comment]
Looking at the compare to try and understand what caused the behavior change: https://github.com/dotnet/roslyn/compare/4c195c3a...9ed4b774

At this point my best guess is that #49765 inadvertently made this warning go away. @cston does this seem like a possibility to you?

**Comment by cston** [comment]
See also https://github.com/dotnet/roslyn/issues/49472.

**Comment by cston** [comment]
> At this point my best guess is that #49765 inadvertently made this warning go away. @cston does this seem like a possibility to you?

It's possible.

**Comment by RikkiGibson** [comment]
In that case I will resolve the bug here. Thank you for filing the bug @madelson and for providing the version info @karataliu.

> Is it a VS issue or SDK issue?

Yes. 😄 We include a compiler in both products, and we try to make sure that we use the same version in each product.
